src/osd/scrubber/scrub_queue.cc

```diff
@@ -0,0 +1,639 @@
+}
+}  // namespace
+
+
+bool ScrubQueue::queue_entries(spg_t pgid, SchedEntry shallow, SchedEntry deep)
+{
+  dout(20) << fmt::format(
+		  "{}: pg[{}]: queuing <{}> & <{}>", __func__, pgid, shallow,
+		  deep)
+	   << dendl;
+  ceph_assert(shallow.pgid == pgid && deep.pgid == pgid);
+
+  if (shallow.urgency == urgency_t::off || deep.urgency == urgency_t::off) {
+    dout(20) << fmt::format(
+		    "{}: pg[{}]: one of the entries is 'off' - not queuing",
+		    __func__, pgid)
+	     << dendl;
+    return false;
+  }
+
+  shallow.is_valid = true;
+  deep.is_valid = true;
+
+  std::unique_lock l{jobs_lock};
+  // now - add the new targets
+  to_scrub.push_back(shallow);
+  to_scrub.push_back(deep);
+  return true;
+}
+
+void ScrubQueue::remove_entry(spg_t pgid, scrub_level_t s_or_d)
+{
+  dout(20) << fmt::format(
+		  "{}: removing {}/{} from the scrub-queue", __func__, pgid,
+		  s_or_d)
+	   << dendl;
+  std::unique_lock l{jobs_lock};
+  auto i = std::find_if(
+      to_scrub.begin(), to_scrub.end(), [pgid, s_or_d](const SchedEntry& t) {
+	return same_key(t, pgid, s_or_d);
+      });
+  if (i != to_scrub.end()) {
+    i->is_valid = false;
+  }
+}
+
+
+void ScrubQueue::cp_and_queue_target(SchedEntry t)
+{
+  dout(20) << fmt::format("{}: restoring {} to the scrub-queue", __func__, t)
+	   << dendl;
+  ceph_assert(t.urgency > urgency_t::off);
+  std::unique_lock l{jobs_lock};
+  t.is_valid = true;
+  to_scrub.push_back(t);
+}
+
+void ScrubQueue::dump_scrubs(ceph::Formatter* f)
+{
+  std::lock_guard lck(jobs_lock);
+  normalize_the_queue();
+
+  f->open_array_section("scrubs");
+  std::for_each(to_scrub.cbegin(), to_scrub.cend(), [&f](const auto& j) {
+    j.dump("sched-target", f);
+  });
+  f->close_section();
+}
+
+
+// ////////////////////////////////////////////////////////////////////////// //
+// initiating a scrub
+
+using ScrubPreconds = Scrub::ScrubPreconds;
+using schedule_result_t = Scrub::schedule_result_t;
+
+void ScrubQueue::sched_scrub(
+    const ceph::common::ConfigProxy& config,
+    bool is_recovery_active)
+{
+  utime_t scrub_tick_time = scrub_clock_now();
+  dout(10) << fmt::format(
+		  "time now:{}, is_recovery_active:{}", scrub_tick_time,
+		  is_recovery_active)
+	   << dendl;
+
+  // do the OSD-wide environment conditions, and the availability of scrub
+  // resources, allow us to start a scrub?
+  auto maybe_env_cond =
+      preconditions_to_scrubbing(config, is_recovery_active, scrub_tick_time);
+  if (!maybe_env_cond) {
+    return;
+  }
+  auto preconds = maybe_env_cond.value();
+
+  std::unique_lock l{jobs_lock};
+
+  // partition and sort the queue
+  if (bool not_empty = normalize_the_queue(); !not_empty) {
+    dout(10) << fmt::format("{}: no eligible scrub targets", __func__) << dendl;
+    return;
+  }
+
+  // pop the first job from the queue, as a candidate
+  auto cand = to_scrub.front();
+  to_scrub.pop_front();
+  l.unlock();
+
+  PgLockWrapper locked_g = osd_service.get_locked_pg(cand.pgid);
```

Comment:
So, this won't work with crimson.  The design here couples the order-a-set-of-items logic with very specific classic logic for finding and locking pgs.  Please decouple those components.